### ISSUE TITLE
[C++][HDFS] Can't get performance improve when increase the thread number of IO thread pool

### ISSUE THREAD

**Issue by youngfn** [apache/arrow#33627]
### Describe the enhancement requested

Hi, I've tested an ExecPlan and the read time is almost 80% of the total time(though I can't give out the exact read time, I've printed out the log in the HDFS read interface, and it shows read ended at almost 80% of total). 
And here is my test results, it tells me the increase of ARROW_IO_THREADS doesn't help to improve performance.
1. total time: 322.502s（OMP_NUM_THREADS=10 ARROW_IO_THREADS=1）     
2. total time: 321.271s（OMP_NUM_THREADS=10 ARROW_IO_THREADS=2）
3. total time: 325.166s（OMP_NUM_THREADS=10 ARROW_IO_THREADS=3）
when I read the code I found no matter how many IO thread Arrow used, **it will only use one HDFS connection**, I draw a thread model I learned from the code(pls point out if I'm wrong).
![thread model](https://user-images.githubusercontent.com/10483852/211994276-a4d54d91-c165-46bb-9f18-2bb059210b87.png)

Also, I've checked the CPU usage when I run the tests, the CPU usage of the read threads is very low, about  15%--22% each. I guess the CPU rate can't increase is because of the output IO upper bound of one HDFS connection.（OMP_NUM_THREADS=10 ARROW_IO_THREADS=3）
![CPU usage](https://user-images.githubusercontent.com/10483852/212008263-a29500b4-e829-4364-a807-91d02dcea93f.jpg)

here are some pieces of my code
```
	int main() {
		arrow::dataset::internal::Initialize();

		// default context will only use one thread
		ExecContext execContext;
		ExecContext exec_context(arrow::default_memory_pool(), ::arrow::internal::GetCpuThreadPool());

		auto&& resultPlan = ExecPlan::Make(&exec_context);
		if (!resultPlan.ok()) {
			WIND_ERR_LOG("Executor run task fail because of make exec plan failed. detail:{}", resultPlan.status().ToString());
			return wind::api::ActionResult{ErrorCode::ERR_COLUMN_ENGINE_INIT};
		}
		std::shared_ptr<ExecPlan> plan = std::move(resultPlan).ValueUnsafe();
		std::vector<Declaration> decls;
		auto actionNode = api::CreateDataFrame(task_.plan, dataFrameMap_, schemas_, *config_, outputPath, decls);

		Declaration::Sequence(decls).AddToPlan(plan.get());

		if (!plan->Validate().ok()) {
			WIND_ERR_LOG("Executor run task fail because of can't get validate exec plan. detail:{}", plan->ToString());
			return wind::api::ActionResult{ErrorCode::ERR_COLUMN_ENGINE_CREATE_EXEC_PLAN};
		}

		plan->StartProducing();
		auto future = plan->finished();

		future.Wait();
	}
	
	std::shared_ptr<ds::Dataset> MakeDataSet() {
		wind::column::MakeFileSystem(schema_.file.path);

		auto format = wind::column::MakeFileFormat(schema_.file.option.csvOption.delimiters[0]);

		auto schema = MakeSchemaAndUpdateFileFormat(format);
		if (!schema) {
			WIND_ERR_LOG("MakeDataSet fail, detail: make schema failed.");
			return nullptr;
		}

		auto fs = wind::column::GetFileSystem();
		if (!fs) {
			WIND_ERR_LOG("MakeDataSet fail, detail: make file system failed.");
			return nullptr;
		}

		auto&& dataset = ds::FileSystemDataset::Make(schema, cp::literal(true), format, fs, MakeFileFragments(format, fs, schema));
		if (!dataset.ok()) {
			WIND_ERR_LOG("MakeDataSet fail, detail:{}", dataset.status().ToString());
			return nullptr;
		}

		return dataset.ValueUnsafe();
	}
	
	std::shared_ptr<ds::ScanOptions> MakeScanOptions(const wind::thrift::ReadLineNode& node,
                                                               const std::shared_ptr<ds::Dataset>& ds) {
		auto options = std::make_shared<ds::ScanOptions>();
		options->use_threads = true;
		arrow::io::IOContext io_context(arrow::default_memory_pool(), ::arrow::internal::GetCpuThreadPool());
		options->io_context = io_context;

		std::vector<std::string> columns;
		for (const auto& column : node.columns) {
			columns.push_back(schema_.fields[column].name);
		}

		// The CsvFileFragment wouldn't benefit from this as much, but
		// can still benefit from skipping conversion of unneeded columns.
		auto resultProject = ds::ProjectionDescr::FromNames(std::move(columns), *(ds->schema()));
		if (!resultProject.ok()) {
			// in this case, convert all columns in table
			options->projection = cp::project({}, {});
			WIND_ERR_LOG("get projectionDescr fail, detail:{}", resultProject.status().ToString());
		} else {
			auto project = std::move(resultProject).ValueUnsafe();
			options->projection = std::move(project.expression);
		}

		if (!node.__isset.filter) return options;

		options->filter = MakeScanFilter(node.filter);
		return options;
	}
```
The ExecPlan for short of my test is like
```
:ConsumingSinkNode{}
  :GroupByNode{keys=["starttime", "cus_area_id", "rat"], aggregates=[
  	hash_sum(intra_rau_req_times),
	....
  ]}
    :ProjectNode{projection=["starttime": get_time(starttime, 86400), cus_area_id, rat...]}
      :FilterNode{filter=((get_time(starttime, 86400) >= 1621699200) and (get_time(starttime, 86400) < 1621785600))}
        :SourceNode{}
```

In conclusion, my questions are these(Arrow version: 10.0.0):
1. Is that add multi-connections to HDFS will help?
2. Am I wrong with using Arrow API so that I can't get improvement when I increase the IO thread pool num?
3. How can I get the performance improvement of reading CSV from HDFS in this case?

### Component(s)

C++

**Comment by westonpace** [comment]
From a glance I don't see anything obvious that would limit the concurrency of HDFS.  It appears the HDFS filesystem will launch a new thread task for each read request.  Do you have multiple CSV files or just one CSV file?

Unfortunately, I do not have an HDFS test environment locally so it is hard for me to test / profile HDFS.

You shouldn't need to do anything to get concurrent I/O.

**Comment by youngfn** [comment]
> 
Do you have multiple CSV files or just one CSV file? -- 
Yes. My test table has 196 files, even though I've set the ARROW_IO_THREADS to 200, but it just runs with 5 reading threads(I print out the log inside hdfs read and count the thread number).
![hdfs read log](https://user-images.githubusercontent.com/10483852/212216366-3638b1c1-9fa8-4b5e-9d51-a85319019996.png)

Unfortunately, I do not have an HDFS test environment locally so it is hard for me to test / profile HDFS. --
It's ok. I have the HDFS env, if you have ideas maybe I can do the test. 
I just want to figure out WHY can't I get the performance improvement by increasing the io thread and what should I do(what parameters should I change in code).

You shouldn't need to do anything to get concurrent I/O. --
Yep, I know it should be. But right now the ARROW_IO_THREADS  and OMP_NUM_THREADS just seem didn't work.

Thank you for replying.

**Comment by westonpace** [comment]
Ah, that number 5 jogged my memory.  Perhaps you need to increase fragment_readahead in ScanOptions?

**Comment by youngfn** [comment]
> 

thx~~~ I will test it right now. 
Btw what about batch_size and batch_readahead, do they affect the thread number?

Also, Do you think the HDFS connection number will affect the reading performance? I mean that each reading thread owns its HDFS connection will read faster than the case all reading threads share one HDFS connection.

**Comment by youngfn** [comment]
> Ah, that number 5 jogged my memory. Perhaps you need to increase fragment_readahead in ScanOptions?

I've changed fragment_readahead to 10(ARROW_IO_THREADS=40) and 40(ARROW_IO_THREADS=200), but it seems nothing changed, still 5 reading threads. And I am sure I've changed the lib after recompiling. So Am I wrong about anything?

**Comment by westonpace** [comment]
No, I don't think you are wrong.  You should be seeing more parallelism.  I am not sure when I will get a chance to fully investigate this however.  I don't see anything in here that would lead threads to share HDFS connections but I don't know anything about how HDFS works.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.
